Update llvm version

paul0403 opened the pull request:
**Context:**
Update llvm version.
I am just trying the latest llvm main branch now.

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

paul0403 commented:
@mehrdad2m just a note that in the pass tutorial, we use the separate `match` and `rewrite`. I think we should update those too. https://docs.pennylane.ai/projects/catalyst/en/stable/dev/transforms.html#writing-transformations-on-catalyst-s-ir

We can do this in a separate PR if this one becomes too big.

paul0403 commented on .github/workflows/build-wheel-linux-x86_64.yaml:
```diff
@@ -245,7 +245,7 @@ jobs:
               -DCMAKE_CXX_FLAGS="-fuse-ld=lld"
 
-        cmake --build $GITHUB_WORKSPACE/enzyme-build --target EnzymeStatic-19
+        cmake --build $GITHUB_WORKSPACE/enzyme-build --target EnzymeStatic-20
```
21

paul0403 commented on mlir/Makefile:
```diff
@@ -119,8 +119,14 @@ mhlo:
+enzyme: PATCH_FILE := $(MK_DIR)/patches/enzyme-nvvm-fabs-intrinsics.patch
 enzyme:
 	@echo "build enzyme"
+	# Patch enzyme's dependency on nvidia fabs llvm intrinsics
```
TODO: patch in CI scripts too

paul0403 commented on .dep-versions:
```diff
@@ -1,16 +1,10 @@
-enzyme=v0.0.149
+mhlo=617a9361d186199480c080c9e8c474a5e30c22d1
+llvm=179d30f8c3fddd3c85056fd2b8e877a4a8513158
+enzyme=v0.0.150 # not 150, just a random number to avoid CI's enzyme cache
```
TODO

paul0403 commented:
As of the previous commit https://github.com/PennyLaneAI/catalyst/pull/1752/commits/32220e5df24200615818e031bc82809f304da864, I can build:
```
mhlo=617a9361d186199480c080c9e8c474a5e30c22d1
llvm=179d30f8c3fddd3c85056fd2b8e877a4a8513158
enzyme=v0.0.180
```

In the above versions, the llvm and mhlo are spit out by the script looking for jax 0.6.0 versions. 
Enzyme is just latest (as of May 22, 2025), with commit db0181320d6e425ee963bd496ed0d8dbb615be18. This is also just enzyme 0.0.180 https://github.com/EnzymeAD/Enzyme/releases/tag/v0.0.180

Patches: 
In mhlo, I remove all shardy-related things.
In enzyme, I remove the nvdia llvm intrinsics.

paul0403 commented:
At the previous commit, only the new buffer deallocation is left. Everything else should be good.

mehrdad2m commented on mlir/Makefile:
```diff
@@ -95,11 +95,35 @@ mhlo: TARGET_FILE := $(MK_DIR)/mlir-hlo/mhlo/transforms/CMakeLists.txt
+
 	# Patch MHLO cmake dependency
 	# TODO: remove once https://github.com/openxla/xla/pull/15446 is merged
 	@if patch --dry-run -p1 -N $(TARGET_FILE) $(PATCH_FILE) > /dev/null 2>&1; then \
```
Do we still need this patch since we have updated mhlo?

paul0403 replied:
https://github.com/PennyLaneAI/catalyst/pull/1752/commits/b04b0592051f183f1bef9680bf74b4da32317426

mehrdad2m commented on mlir/lib/Driver/Pipelines.cpp:
```diff
@@ -73,26 +73,23 @@ void createBufferizationPipeline(OpPassManager &pm)
     pm.addNestedPass<mlir::func::FuncOp>(mlir::bufferization::createBufferLoopHoistingPass());
     pm.addNestedPass<mlir::func::FuncOp>(mlir::bufferization::createPromoteBuffersToStackPass());
-    pm.addNestedPass<mlir::func::FuncOp>(mlir::bufferization::createBufferDeallocationPass());
+    // TODO: what's the cpp for buffer-deallocation-pipeline???
```
you can probably use [this function](https://github.com/llvm/llvm-project/blob/02916a432ca6465e2e45f67be94f1c89c9cd425d/mlir/lib/Dialect/Bufferization/Pipelines/BufferizationPipelines.cpp#L21-L36) like:

```
mlir::bufferization::BufferDeallocationPipelineOptions bufferDeallocOptions;
mlir::bufferization::buildBufferDeallocationPipeline(pm, bufferDeallocOptions);
```

paul0403 replied:
https://github.com/PennyLaneAI/catalyst/pull/1752/commits/e1d3c962584ff3da639955456835d00ef4082165

mehrdad2m reviewed (commented):
🚀🚀 🚀

paul0403 commented on mlir/test/Quantum/AllocationTest.mlir:
```diff
@@ -12,25 +12,27 @@
+func.func @existing_deallocation() -> i64 {
     // CHECK: [[reg:%.+]] = quantum.alloc
     %0 = quantum.alloc( 1) : !quantum.reg
+    %user = builtin.unrealized_conversion_cast %0 : !quantum.reg to i64
```
I finished the description writeup. See point 3 of lit test updates.

paul0403 commented on mlir/lib/Quantum/Transforms/BufferizableOpInterfaceImpl.cpp:
```diff
@@ -177,7 +177,7 @@ struct SampleOpInterface
     }
 
-    bool bufferizesToAllocation(Operation *op, Value value) const { return true; }
+    // bool bufferizesToAllocation(Operation *op, Value value) const { return true; }
```
It should be removed, thanks!

paul0403 replied:
See the last item in point 14 of the description:

> The migration guide also suggests that bufferizable ops no longer implement the bufferizesToAllocation method, so we remove them.

dime10 commented on .github/workflows/check-catalyst.yaml:
```diff
@@ -146,7 +146,7 @@ jobs:
         sudo apt-get install -y python3 python3-pip cmake ninja-build clang lld
         python3 --version | grep ${{ needs.constants.outputs.primary_python_version }}
-        python3 -m pip install numpy pybind11
+        python3 -m pip install numpy pybind11 nanobind==2.4
```
Definitely should put in in requirements if we now need this :)

Oh nice, so they have moved away from pybind as well? I think that means we could add back the extension modules to our bindings if we ever need to, since we removed them in order to support the ABI3 wheels with nanobind.

mehrdad2m replied:
yes I think so. They still have pybind11 as a requirement, but this [PR](https://github.com/llvm/llvm-project/pull/120473/files) essentially ports all the Python core code to nanobind.

dime10 commented on frontend/catalyst/pipelines.py:
```diff
@@ -240,7 +240,7 @@ def get_bufferization_stage(options: CompileOptions) -> List[str]:
         "func.func(buffer-loop-hoisting)",
         "func.func(promote-buffers-to-stack)",
-        "func.func(buffer-deallocation)",
+        "buffer-deallocation-pipeline",
```
This is potentially a big change that might have been worth putting in its own PR, like we did for the bufferization, independent of a version upgrade. First and foremost, it would be good to know whether the deallocation behaviour has changed at all compared to the previous pass.

paul0403 replied:
I agree, migration of buffer deallocation is a change that should be in its own PR.

However, the old `--buffer-deallocation` is completely removed. https://github.com/llvm/llvm-project/pull/126366
Therefore we must address this, or we can't finish the llvm update.

paul0403 replied:
The deallocation behavior changed a lot: https://mlir.llvm.org/docs/OwnershipBasedBufferDeallocation/

... which is also why more time should be spent here. But I am unsure, given how much time we have spent on this update.

paul0403 replied:
The alternative strategy is to again, patch back the old buffer deallocation pass, and leave this as a TODO for next quarter's update, or just even a regular item. This would also mean we don't need to deal with the memory effects in this PR. 

Although I really want to avoid more patches :melting_face:

dime10 commented on mlir/Makefile:
```diff
@@ -95,11 +95,35 @@ mhlo: TARGET_FILE := $(MK_DIR)/mlir-hlo/mhlo/transforms/CMakeLists.txt
+	touch $(MK_DIR)/mlir-hlo/mhlo/transforms/legalize_to_standard/legalize_to_standard_patterns.td
+	@if patch -p0 --dry-run -N --verbose < $(MK_DIR)/patches/mhlo-add-back-necessary-passes.patch > /dev/null 2>&1; then \
+		patch -p0 < $(MK_DIR)/patches/mhlo-add-back-necessary-passes.patch; \
+	fi
```
These are the patches? 😬 That's worse than I thought, I didn't know they just removed everything. Should be high priority on our backlog to address this situation.

dime10 commented on mlir/Makefile:
```diff
@@ -95,11 +95,35 @@ mhlo: TARGET_FILE := $(MK_DIR)/mlir-hlo/mhlo/transforms/CMakeLists.txt
+	#rm -rf $(MK_DIR)/mlir-hlo/mhlo/transforms/legalize_control_flow
+	#rm -rf $(MK_DIR)/mlir-hlo/mhlo/transforms/legalize_sort
+	#rm -rf $(MK_DIR)/mlir-hlo/mhlo/transforms/legalize_to_standard
+
```
Are you deleting this? 
```suggestion
```

dime10 commented on mlir/include/Catalyst/IR/CatalystOps.td:
```diff
@@ -41,7 +41,8 @@ def ListDeallocOp : Catalyst_Op<"list_dealloc"> {
                         "list", "value",
-                        "mlir::cast<ArrayListType>($_self).getElementType()">]> {
+                        "mlir::cast<ArrayListType>($_self).getElementType()">,
+         MemoryEffects<[MemWrite, MemAlloc]>]> {
```
I was gonna ask why the alloc is here, but I guess you're thinking about the case when the arraylist needs to be extended. In that case, wouldn't you also want a `MemFree`, since resizing involves alloc - copy - free?

The problem I'm thinking about though is that we have to distinguish between the semantics of an operation (its conceptual definition) and its _implementation_. With the reasoning above, you might start attaching memory semantics to all the tensor ops as well because they are implemented as memory reads/writes eventually, but that obviously that defeats the purpose of defining tensor semantics to begin with. 

If this op carries allocation semantics, where is the allocated resource exactly? The op doesn't have any results.

I think a more accurate model of this op is just as a "write" to the list argument. We could even define lists as their own resource separate from the built-in memory resource, to distinguish the actions of these ops from those of memref ops, but I don't actually know what practical difference that would make 🤔

dime10 commented on mlir/include/Catalyst/IR/CatalystOps.td:
```diff
@@ -71,7 +73,7 @@ def ListLoadDataOp : Catalyst_Op<"list_load_data"> {
 }
 
-def PrintOp : Catalyst_Op<"print"> {
+def PrintOp : Catalyst_Op<"print", [MemoryEffects<[MemRead, MemAlloc]>]> {
```
Why is this an alloc?

paul0403 replied:
Prints will get removed if it's just memread since it has no results and thus no users. 

Between write and alloc, I thought alloc was more appropriate because sometimes it needs to alloc the space for the words it's printing.

dime10 replied:
I think write is more appropriate, since print can be modelled as writing to some IO buffer. We could put a write to the runtime resource here for example.

dime10 commented on mlir/include/Catalyst/IR/CatalystOps.td:
```diff
@@ -81,7 +83,7 @@ def PrintOp : Catalyst_Op<"print"> {
 }
 
-def AssertionOp : Catalyst_Op<"assert"> {
+def AssertionOp : Catalyst_Op<"assert", [MemoryEffects<[MemWrite]>]> {
```
Assertions are interesting, I think a "write" is the right choice to some abstract computer state.

paul0403 replied:
Just a note that practically, ops without results (like assert here) and with no memory effects or read-only effects get removed. This is why I added memwrite here. My thinking was "it needs to write the assertion result somewhere". But I don't like this line of reasoning. Assertion being read-only feels natural to me but then they would get removed.

paul0403 replied:
Hmm, makes sense, I guess assert does change the "computer state" depending on whether it's true or false. All good then. 

I'm resolving this as there's no change needed to the code.

dime10 commented on mlir/include/Gradient/IR/GradientOps.td:
I don't see a reason to attach alloc/free semantics to these ops, none of them are explicitly modelled to mean those things. Read/Write to the memref arguments is correct, although what's interesting is that in their "tensor" version these ops would not have such semantics and in fact be side-effect free. I think a way to model this (other than defining two sets of ops) might be to attach the effects to the arguments/results themselves, instead of the operation.

dime10 commented on mlir/include/Quantum/IR/QuantumDialect.td:
```diff
@@ -72,4 +73,24 @@ def ResultType : Quantum_Type<"Result", "res"> {
+    // This resource represents a generic piece of quantum memory. It can be used to
+    // model the resource bahavior of quantum operations, in order to help make decisions
+    // during various IR analyses.
+    // For example, a gate operation will both read from and write to quantum memory.
```
This is not true since we use value-semantics for our gates, although it is an alternative way to model gates.

dime10 replied:
Yeah it was unused previously, we probably introduced when trying out the memory modelling for registers.

paul0403 replied:
Yes, I just realized that these memory effects are only IR side effects used for hints during IR transformations, e.g. ops ordering, ops removal/cse, ..., and maybe shouldn't be taken too literally?

According to [side effect docs](https://mlir.llvm.org/docs/Rationale/SideEffectsAndSpeculation/):

> Many MLIR operations don’t exhibit any behavior other than consuming and producing SSA values. These operations can be reordered with other operations as long as they obey SSA dominance requirements and can be eliminated or even introduced (e.g. for [rematerialization](https://en.wikipedia.org/wiki/Rematerialization)) as needed.
However, a subset of MLIR operations have implicit behavior than isn’t reflected in their SSA data-flow semantics. These operations need special handing, and cannot be reordered, eliminated or introduced without additional analysis.

So it sounds like these side effects are additional hints on top of the usual information available from SSA data flow.

So yes for the gate ops we have it would indeed mean the side effects are not necessary, as we chain gates up (and lock down their order) with explicit SSA semantics already. 

I will remove this line in the description.

paul0403 replied:
https://github.com/PennyLaneAI/catalyst/pull/1752/commits/0f54789e513f9cc449642cdc77f4ff0d4c206df8

dime10 commented on mlir/include/Quantum/IR/QuantumOps.td:
```diff
@@ -63,23 +60,23 @@ def NamedObservableAttr : EnumAttr<QuantumDialect, NamedObservable, "named_obser
 class Quantum_Op<string mnemonic, list<Trait> traits = []> : Op<QuantumDialect, mnemonic, traits>;
 
-def InitializeOp : Quantum_Op<"init"> {
+def InitializeOp : Quantum_Op<"init", [MemoryEffects<[MemAlloc<QuantumMemory>]>]> {
```
I think the runtime should be a different resource than quantum memory (it's a classical resource), and so should quantum devices (somewhat quantum, but not really quantum memory i.e. qubits) 🤔

erick-xanadu replied:
There should be a runtime resource :sweat_smile:

dime10 commented on mlir/include/Quantum/IR/QuantumOps.td:
```diff
@@ -172,7 +169,7 @@ def DeallocQubitOp : Memory_Op<"dealloc_qb"> {
 }
 
-def DeallocOp : Memory_Op<"dealloc"> {
+def DeallocOp : Memory_Op<"dealloc", [MemoryEffects<[MemFree<QuantumMemory>]>]> {
```
The asymmetry between the alloc and dealloc really bothers me, but I guess it was one way achieve what we wanted (which was being able to DCE allocations if they are unused, but maintain deallocs as a valid use that should never be deleted).

dime10 commented on mlir/include/Quantum/IR/QuantumOps.td:
```diff
@@ -341,7 +338,8 @@ class UnitaryGate_Op<string mnemonic, list<Trait> traits = []> :
 }
 
-def SetStateOp : Gate_Op<"set_state", []> {
+def SetStateOp : Gate_Op<"set_state",
```
All gates should be pure (due the qubit value semantics).

Although, do the memrefs need a read?

dime10 reviewed (changes requested):
Leaving some initial comments on the first part of the PR :)

dime10 commented on mlir/test/Quantum/AllocationTest.mlir:
```diff
@@ -12,25 +12,27 @@
     // CHECK: quantum.dealloc [[reg]]
     quantum.dealloc %0 : !quantum.reg
-    return
+    return %user : i64
```
Just returning %0 directly should be sufficient to preserve the instructions from DCE 😅

paul0403 replied:
https://github.com/PennyLaneAI/catalyst/pull/1752/commits/ab3103496df4ea7c37b46d72d748d19d800e5101

dime10 commented on mlir/lib/Quantum/Transforms/emit_catalyst_pyface.cpp:
```diff
@@ -212,16 +212,21 @@ struct EmitCatalystPyInterfacePass
         config.strictMode = GreedyRewriteStrictness::ExistingOps;
         config.enableRegionSimplification = mlir::GreedySimplifyRegionLevel::Disabled;
         config.maxIterations = 1;
+        // Update to the following lines the next time we update llvm
```
```suggestion
        // TODO: Update to the following lines the next time we update llvm
```

dime10 commented on mlir/lib/Catalyst/Transforms/InlineNestedModules.cpp:
```diff
@@ -410,7 +411,9 @@ struct InlineNestedSymbolTablePass : PassWrapper<InlineNestedSymbolTablePass, Op
         config.strictMode = GreedyRewriteStrictness::ExistingOps;
         config.enableRegionSimplification = mlir::GreedySimplifyRegionLevel::Disabled;
-
+        // Update to the following lines the next time we update llvm
```
```suggestion
        // TODO: Update to the following lines the next time we update llvm
```

dime10 commented on mlir/lib/Catalyst/Transforms/InlineNestedModules.cpp:
```diff
@@ -379,18 +379,19 @@ struct AnnotateWithFullyQualifiedNamePass
         GreedyRewriteConfig config;
         config.strictMode = GreedyRewriteStrictness::ExistingOps;
         config.enableRegionSimplification = mlir::GreedySimplifyRegionLevel::Disabled;
+        // Update to the following lines the next time we update llvm
```
```suggestion
        // TODO: Update to the following lines the next time we update llvm
```

dime10 commented on mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:
```diff
@@ -926,8 +920,11 @@ struct AddExceptionHandlingPass : impl::AddExceptionHandlingPassBase<AddExceptio
         GreedyRewriteConfig config;
         config.strictMode = GreedyRewriteStrictness::ExistingOps;
         config.enableRegionSimplification = mlir::GreedySimplifyRegionLevel::Disabled;
+        // Update to the following lines the next time we update llvm
```
```suggestion
        // TODO: Update to the following lines the next time we update llvm
```

dime10 commented on mlir/lib/Gradient/Transforms/GradMethods/Adjoint.cpp:
```diff
@@ -82,22 +90,43 @@ func::FuncOp AdjointLowering::discardAndReturnReg(PatternRewriter &rewriter, Loc
             rewriter.create<func::FuncOp>(loc, fnName, fnType, visibility, nullptr, nullptr);
-        // clone the body.
+        // Clone the body.
+        // TODO: we do not support submodules, i.e. our gradient functions
+        // will have just one block.
+        assert(callee.getBody().hasOneBlock() &&
+               "Gradients with quantum submodules are not supported");
```
subroutines? 😅 
```suggestion
        // TODO: we do not support subroutines, i.e. our gradient functions
        // will have just one block.
        assert(callee.getBody().hasOneBlock() &&
               "Gradients with quantum subroutines are not supported");
```

Although this is odd, if you want to check that there are no subroutines, you should verify that there are no call ops (or call-like ops) in the body. Checking for a single block just asserts that there is no unstructured control flow.

dime10 replied:
Oh right, I remember now. I think you wanted to check that there is only a single block so that we can assume there is only one return statement? Which I think is true anyway, as we only deal with structured control flow until going to llvm, so that makes sense to assert.

dime10 commented on mlir/lib/Gradient/Transforms/GradMethods/Adjoint.cpp:
```diff
@@ -68,10 +65,21 @@ func::FuncOp AdjointLowering::discardAndReturnReg(PatternRewriter &rewriter, Loc
+    Type f64Type = rewriter.getF64Type();
+    SmallVector<Type> retTypes{qregType};
+    auto expvalOps = callee.getOps<quantum::ExpvalOp>();
+    std::for_each(expvalOps.begin(), expvalOps.end(),
```
We should assert that our assumption is true, i.e. that there are no other measurement processes besides expval.

dime10 commented on mlir/lib/Gradient/Transforms/GradMethods/Adjoint.cpp:
```diff
@@ -82,22 +90,43 @@ func::FuncOp AdjointLowering::discardAndReturnReg(PatternRewriter &rewriter, Loc
+        SmallVector<Value> returnVals;
+
+        // Fine for now: only one block in body so only one quantum.dealloc
+        auto localDealloc = *unallocFn.getOps<quantum::DeallocOp>().begin();
```
Grab the terminator directly, it's faster faster than iterating to find ops of a particular type :)

dime10 commented on mlir/lib/Gradient/Transforms/GradMethods/Adjoint.cpp:
```diff
@@ -82,22 +90,43 @@ func::FuncOp AdjointLowering::discardAndReturnReg(PatternRewriter &rewriter, Loc
+
+        // We also need to return the expval op so it won't be dead-code-eliminated
+        // TODO: we only support grad on expval op for now
+        for (auto expvalOp : unallocFn.getOps<quantum::ExpvalOp>()) {
```
Wait if we're already iterating through the expval ops above (to get the types), can't you do the collection there too?

dime10 replied:
My bad 👍

paul0403 replied:
https://github.com/PennyLaneAI/catalyst/pull/1752/commits/4cc38e48ee385d2f2836704737896bc9c902cbf6

paul0403 replied:
https://github.com/PennyLaneAI/catalyst/pull/1752/commits/034098e969957399d40cb3fe1d1cf77d291d417a

dime10 commented on mlir/lib/Gradient/Transforms/GradMethods/Adjoint.cpp:
```diff
@@ -82,22 +90,43 @@ func::FuncOp AdjointLowering::discardAndReturnReg(PatternRewriter &rewriter, Loc
+            returnVals.push_back(expvalOp);
+        }
+
+        // Erase the device release.
+        // Unfortunately erasing on the fly within the getOps() causes the
+        // "erasing while iterating" segfault
+        // So we need a separate, completely materialized worklist.
```
With the `.walk` method from before it was possible though?

dime10 commented on mlir/lib/Gradient/Transforms/GradMethods/Adjoint.cpp:
```diff
@@ -82,22 +90,43 @@ func::FuncOp AdjointLowering::discardAndReturnReg(PatternRewriter &rewriter, Loc
+
+        // Create the return
+        // Again, assume just one block for now
+        (*unallocFn.getOps<func::ReturnOp>().begin())->setOperands(returnVals);
```
Just generally it's good to be mindful of how often you're iterating over a possibly large piece of IR, as that can start to add up, especially once a function is called by other functions which can quickly become quadratic (not necessarily the case here, just in general).

dime10 reviewed (commented):
Great work, especially the detailed PR description 🥇

paul0403 commented:
I split the new buffer dealloc migration into its own PR #1778 and added back the old buffer dealloc.

I am therefore closing all comments regarding the memory effects. We will properly revisit them the next time...

paul0403 commented on mlir/lib/Catalyst/Transforms/BufferizableOpInterfaceImpl.cpp:
```diff
@@ -75,8 +75,6 @@ struct PrintOpInterface
 struct CustomCallOpInterface
     : public bufferization::BufferizableOpInterface::ExternalModel<CustomCallOpInterface,
                                                                    CustomCallOp> {
-    bool bufferizesToAllocation(Operation *op, Value value) const { return true; }
```
I guess we should add these `bufferizesToAllocation` methods back now that we are not migrating this time.

paul0403 commented on mlir/Makefile:
```diff
@@ -95,11 +95,24 @@ mhlo: TARGET_FILE := $(MK_DIR)/mlir-hlo/mhlo/transforms/CMakeLists.txt
 	@echo "build MLIR-HLO"
+	# Clean mhlo submodule for git patches
+	make clean-mhlo
+
```
I found the magic incantation that dry-runs for `git apply`: 
https://git-scm.com/docs/git-apply#Documentation/git-apply.txt---check

https://github.com/PennyLaneAI/catalyst/pull/1752/commits/e9fbb2d4e8cac0f68029a0392bdd080c530e4a0c

paul0403 replied:
(so I removed this clean)

paul0403 commented on mlir/Makefile:
```diff
@@ -204,6 +206,7 @@ clean-llvm:
 clean-mhlo:
 	@echo "clean HLO dialect build files"
 	rm -rf $(MHLO_BUILD_DIR)
+	cd mlir-hlo; git clean -fd; git checkout .; \
```
Good catch :sweat_smile:

paul0403 replied:
https://github.com/PennyLaneAI/catalyst/pull/1752/commits/8ff300f79d6d97f567d67e05bb35ba0fbdea259c

paul0403 commented on requirements.txt:
```diff
@@ -5,7 +5,7 @@ pip>=22.3
 # Do not allow NumPy 2.0.0 due to a bug with their C API that blocks the usage of the Stable ABI;
 # this bug was fixed in 2.0.1 (https://github.com/numpy/numpy/pull/26995)
-nanobind
+nanobind==2.4
 numpy!=2.0.0
```
@mehrdad2m has more context on this, is 2.4 required exactly?

dime10 replied:
Actually 3 hasn't even been released yet, so I'm guessing they do this as a precaution. Since we don't really follow this practice, I would say just `nanobind>=2.4` is also fine.

paul0403 replied:
https://github.com/PennyLaneAI/catalyst/pull/1752/commits/38cd1291ffcacb932d1658bf51b11d2a20b3e412 updated to >=2.4

paul0403 replied:
added a comment explaining https://github.com/PennyLaneAI/catalyst/pull/1752/commits/0084fe745b662f2a3b5c0df565632f58bd91bb4c

paul0403 commented on mlir/test/Gradient/AdjointTest.mlir:
I was checking for exactly one expval op.

But I now realize multiple expval is ok :sweat_smile: It will just be a jacobian. I will change them back.

paul0403 replied:
I updated so that (1) number of expvals is not constrained, (2) we check for there's just one device release, and (3) revert these test updates

https://github.com/PennyLaneAI/catalyst/pull/1752/commits/1240c6a0285339fc63d1aa659c996e23af1700fc

paul0403 commented on mlir/include/Quantum/IR/QuantumOps.td:
```diff
@@ -567,7 +564,7 @@ def GlobalPhaseOp : UnitaryGate_Op<"gphase", [DifferentiableGate, AttrSizedOpera
 
 def MultiRZOp : UnitaryGate_Op<"multirz", [DifferentiableGate, NoMemoryEffect,
-                                           AttrSizedOperandSegments, AttrSizedResultSegments]> {
+                            AttrSizedOperandSegments, AttrSizedResultSegments]> {
```
This is a relic of the memory effect changes. I'll indent them back.

paul0403 commented on mlir/lib/Gradient/Transforms/GradMethods/Adjoint.cpp:
```diff
@@ -47,15 +44,44 @@ void AdjointLowering::rewrite(func::FuncOp op, PatternRewriter &rewriter) const
+                continue;
+            }
+            else {
+                callee.emitOpError() << "Adjoint gradient is only supported on expval measurements";
```
Ah yes, returning right away here does make sense. I will add this.

paul0403 replied:
https://github.com/PennyLaneAI/catalyst/pull/1752/commits/13853d529d759797ed860e1ed1f7d68acd257161

paul0403 commented on mlir/lib/Gradient/Transforms/GradMethods/Adjoint.cpp:
```diff
@@ -67,11 +93,18 @@ func::FuncOp AdjointLowering::discardAndReturnReg(PatternRewriter &rewriter, Loc
+    if (numExpvals != 1) {
+        callee.emitOpError() << "Invalid number of expval ops: " << numExpvals;
+        return callee;
+    }
```
I re-enabled multiple expvals. See the other comment.

paul0403 commented on mlir/lib/Catalyst/Transforms/TBAAPatterns.cpp:
```diff
@@ -12,20 +12,77 @@
+    bool memIsFuncArg = (std::find(funcArgs.begin(), funcArgs.end(), mem) != funcArgs.end());
+    if (!memIsFuncArg) {
+        return false;
+    }
```
https://github.com/PennyLaneAI/catalyst/pull/1752/commits/b6833bf60b6c5a5059c6b8accb3079ccbcaee89d

paul0403 commented on mlir/lib/Catalyst/Transforms/TBAAPatterns.cpp:
```diff
@@ -12,20 +12,77 @@
+    if (prevOp) {
+        return isFromExtractAlignedPointerAsIndexOp(prevOp);
+    }
+    return false;
```
https://github.com/PennyLaneAI/catalyst/pull/1752/commits/f7455ddec9458cdab61207d8c932b2939d76cfaf

dime10 commented:
I'm investigating issue #2420, and I'm pretty sure this PR introduced the bug.

Looking at this change specifically, I think it misinterpreted the restriction in place on the adjoint method:
>13. The new pattern applicator (`applyPatternsGreedily`) is more
aggressive in dead code elimination,
and is eliminating dead `Value`s in the adjoint gradient method.
The `nodealloc` function we generate for adjoint gradient lowering used
to only return the qreg, not the expval result. This causes the expval
op to be eliminated since it has no users.
This further causes wrong gradient results, since the entire program,
all ops included (regardless
of dead or not), impacts the gradient through chain rule.
To avoid this, we return the expval result as well.
In doing this, **we implicitly assume that differentiated qnodes can only
return expval.
Although this assumption is true and also restricted by frontend,**
ideally we should not have it hard coded.
We leave this as a TODO for a future feature.

Specifically, we had:
```cpp
        // The callee of the adjoint op must return as a single result the quantum register.
        func::FuncOp callee =
            SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, op.getCalleeAttr());
        assert(callee && callee.getNumResults() == 1 && "invalid qfunc symbol in adjoint op");
```
where the generated function only returned the register, but this did not restrict how many QNode MP results could be produced. Now the check reads:
```cpp
        // The callee of the adjoint op must return 2 results: the quantum register and the expval.
        func::FuncOp callee =
            SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, op.getCalleeAttr());
        assert(callee && callee.getNumResults() == 2 && "invalid qfunc symbol in adjoint op");
```

In the frontend we only ever had this verification:
```py
    if f.diff_method == "adjoint" and any(prim not in [expval_p] for prim in return_ops):
        raise DifferentiableCompileError(
            "The adjoint method can only be used for QNodes which return qml.expval."
        )
```
It only checks that all results are expvals, but not how many there are.

Looking into whether we can reverse this change / fix the problem that prompted this change another way.
